dbuf-core/src/codegen/codegen.rs

```diff
@@ -0,0 +1,241 @@
+use core::panic;
+use std::iter;
+
+use crate::ast::elaborated::*;
+use pretty::BoxDoc;
+
+type Str = String; // Temporary
+const NEST_UNIT: isize = 4;
+
+#[derive(Clone)]
+struct NamedConstructor<Name, Constructor> {
+    name: Name,
+    constructor: Constructor,
+}
+
+struct RustType(String);
+
+// everything is needed to know about arbitrary field
+struct Field {
+    // name of the field in dependobuf
+    name: Str,
+
+    // dependobuf type
+    ty: TypeExpression<Str>,
+
+    // associated name when is part of some struct
+    private_name: Str,
+
+    // associated type when is part of some struct
+    private_ty: RustType,
+}
+
+// TODO: change panics to regular error handling
+// TODO: write adequate doc
+// TODO: add rust pretty printer in order to simplify most of the generation here
+// TODO: do something with lifetimes, they are shouldn't be needed
+
+// This is sketch implementation.
+// Final implementation will use own ast, because elaborated is not quite what codegen needs.
+// For example types in elaborated module are ordered in topological order over dependencies
+// however when generating, codegen is not enforced to place struct in such order.
+// For now however order is now even a concern because everything is passed inside Boxes.
+struct ModuleCodegen {
+    module: Module<Str>,
+}
+
+impl ModuleCodegen {
+    pub fn generate<'a>(&'a self) -> BoxDoc<'a> {
+        let types = self
+            .module
+            .types
+            .iter()
+            .map(|(name, body)| self.generate_type(name, body));
+        BoxDoc::text("type Box<T> = std::boxed::Box<T>;")
+            .append(BoxDoc::line())
+            .append(BoxDoc::intersperse(types, BoxDoc::line()))
+    }
+
+    fn generate_type<'a>(&'a self, name: &'a Str, body: &'a Type<Str>) -> BoxDoc<'a> {
```

Comment:
In this and following functions, a `self` argument is actually unused so I'd make them just standalone functions